src/NuGet.Services.AzureSearch/Catalog2AzureSearch/AzureSearchCollectorLogic.cs

```diff
@@ -0,0 +1,220 @@
+            IEnumerable<CatalogCommitItem> items)
+        {
+            // Generate the work by taking the latest catalog leaf per package identity then grouping the remaining
+            // items by package ID.
+            var workEnumerable = items
+                .GroupBy(x => x.PackageIdentity)
+                .Select(x => GetLatest(x))
+                .GroupBy(x => x.PackageIdentity.Id, StringComparer.OrdinalIgnoreCase)
+                .Select(x => new Work(x.Key, x.ToList()));
+            var allWork = new ConcurrentBag<Work>(workEnumerable);
+
+            // Process the package ID groups in parallel, collecting all index actions for later.
+            var output = new ConcurrentBag<IdAndValue<IndexActions>>();
+            var tasks = Enumerable
+                .Range(0, _options.Value.MaxConcurrentBatches)
+                .Select(async x =>
+                {
+                    await Task.Yield();
+                    while (allWork.TryTake(out var work))
+                    {
+                        var entryToLeaf = await GetEntryToLeaf(work.Entries);
+                        var indexActions = await GetPackageIdIndexActionsAsync(work.Entries, entryToLeaf);
+                        output.Add(new IdAndValue<IndexActions>(work.PackageId, indexActions));
+                    }
+                })
+                .ToList();
+            await Task.WhenAll(tasks);
+
+            return output;
+        }
+
+        private CatalogCommitItem GetLatest(IEnumerable<CatalogCommitItem> entries)
+        {
+            CatalogCommitItem max = null;
+            foreach (var entry in entries)
+            {
+                if (max == null)
+                {
+                    max = entry;
+                    continue;
+                }
+
+                Guard.Assert(
+                    StringComparer.OrdinalIgnoreCase.Equals(max.PackageIdentity, entry.PackageIdentity),
+                    "The entries compared should have the same package ID.");
+
+                if (entry.CommitTimeStamp > max.CommitTimeStamp)
+                {
+                    max = entry;
+                }
+                else if (entry.CommitTimeStamp == max.CommitTimeStamp)
+                {
+                    const string message = "There are multiple catalog leaves for a single package at one time.";
+                    _logger.LogError(
+                        message + " ID: {PackageId}, version: {PackageVersion}, commit timestamp: {CommitTimestamp:O}",
+                        entry.PackageIdentity.Id,
+                        entry.PackageIdentity.Version.ToFullString(),
+                        entry.CommitTimeStamp);
+                    throw new InvalidOperationException(message);
+                }
+            }
+
+            return max;
+        }
+
+        private async Task<IReadOnlyDictionary<CatalogCommitItem, PackageDetailsCatalogLeaf>> GetEntryToLeaf(
```

Comment:
>GetEntryToLeaf [](start = 94, length = 14)

This is called by a `Work` item in `ProcessWorkAsync`. Won't this mean that you'll be processing up to `MaxConcurrentBatches * MaxConcurrentBatches` package details leaves at once? Could you instead build up this dictionary before you start processing `Work` items?

Reply:
Yeah, nice catch. I originally had this but broke it during a collector refactoring 😓